usr/lib/soft_stdll/soft_specific.c

```diff
@@ -3051,3 +3075,540 @@ CK_RV token_specific_aes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
 
     return rv;
 }
+
+#ifndef NO_EC
+
+static CK_RV make_ec_key_from_params(const CK_BYTE *params, CK_ULONG params_len,
+                                     EC_KEY **key)
+{
+    const unsigned char *oid;
+    ASN1_OBJECT *obj = NULL;
+    EC_KEY *ec_key = NULL;
+    int nid;
+    CK_RV rc = CKR_OK;
+
+    oid = params;
+    obj = d2i_ASN1_OBJECT(NULL, &oid, params_len);
+    if (obj == NULL) {
+        TRACE_ERROR("curve not supported by OpenSSL.\n");
+        rc = CKR_CURVE_NOT_SUPPORTED;
+        goto out;
+    }
+
+    nid = OBJ_obj2nid(obj);
+    if (nid == NID_undef) {
+        TRACE_ERROR("curve not supported by OpenSSL.\n");
+        rc = CKR_CURVE_NOT_SUPPORTED;
+        goto out;
+    }
+
+    ec_key = EC_KEY_new_by_curve_name(nid);
+    if (ec_key == NULL) {
+       TRACE_ERROR("curve not supported by OpenSSL.\n");
+       rc = CKR_CURVE_NOT_SUPPORTED;
+       goto out;
+    }
+
+out:
+    if (obj != NULL)
+        ASN1_OBJECT_free(obj);
+
+    if (rc != CKR_OK) {
+        if (ec_key != NULL)
+            EC_KEY_free(ec_key);
+
+        return rc;
+    }
+
+    *key = ec_key;
+
+    return CKR_OK;
+}
+
+static CK_RV fill_ec_key_from_pubkey(EC_KEY *ec_key, const CK_BYTE *data,
+                                     CK_ULONG data_len)
+{
+    CK_BYTE *ecpoint = NULL;
+    CK_ULONG ecpoint_len, field_len, privlen, padlen;
+    CK_BYTE form, *temp = NULL;
+    CK_RV rc;
+
+    /* CKA_EC_POINT contains the EC point as OCTET STRING */
+    rc = ber_decode_OCTET_STRING((CK_BYTE *)data, &ecpoint, &ecpoint_len,
+                                 &field_len);
+    if (rc != CKR_OK || field_len != data_len) {
+        TRACE_DEVEL("ber_decode_OCTET_STRING failed\n");
+        rc = CKR_ATTRIBUTE_VALUE_INVALID;
+        goto out;
+    }
+
+    /* Check for public key without format byte */
+    privlen = (EC_GROUP_order_bits(EC_KEY_get0_group(ec_key)) + 7) / 8;
+    form  = ecpoint[0] & ~0x01;
+    if (ecpoint_len <= 2 * privlen &&
+        form != POINT_CONVERSION_COMPRESSED &&
+        form != POINT_CONVERSION_UNCOMPRESSED &&
+        form != POINT_CONVERSION_HYBRID) {
+        temp = malloc(1 + 2 * privlen);
+        if (temp == NULL) {
+            rc = CKR_HOST_MEMORY;
+            goto out;
+        }
+
+        padlen = 2 * privlen - ecpoint_len;
+        temp[0] = POINT_CONVERSION_UNCOMPRESSED;
+        memset(temp + 1, 0, padlen);
+        memcpy(temp + 1 + padlen, ecpoint, ecpoint_len);
+
+        ecpoint = temp;
+        ecpoint_len = 1 + 2 * privlen;
+    }
+
+    if (!EC_KEY_oct2key(ec_key, ecpoint, ecpoint_len, NULL)) {
+        TRACE_ERROR("EC_KEY_oct2key failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
```

Comment:
Same as above, goto out is not necessary here, but I would even prefer leaving it as is ...